### ISSUE TITLE
Clarify [[Call]] + [[Construct]] for named (and possibly non-named) constructors

### ISSUE THREAD

**Issue by domenic** [heycam/webidl#278]
We should state that named constructors are [built-in function objects](https://tc39.github.io/ecma262/#sec-built-in-function-objects) whose behavior (both [[Call]] and [[Construct]] is specified as the steps currently given. That is, we need to trigger the clause in the definition of built-in function objects that says

> The behaviour specified for each built-in function via algorithm steps or other means is the specification of the function body behaviour for both [[Call]] and [[Construct]] invocations of the function. 

The spec currently says "When evaluating the function object F" for non-[NamedConstructor] interface objects, which is a reasonable way of specifying that behavior.

Additionally the ES spec says

> Built-in function objects that are not identified as constructors do not implement the [[Construct]] internal method unless otherwise specified in the description of a particular function.

So we should be sure to say that they are constructors. Maybe it's better to just be explicit and say that the steps are for both [[Construct]] and [[Call]] in both cases.

**Comment by tobie** [comment]
It seems named constructors throw when called without a NewTarget. At least `Image` does in both Chrome and Firefox.

**Comment by tobie** [comment]
See https://github.com/heycam/webidl/issues/275

**Comment by tobie** [comment]
I think that's implicit in the abstract operations defined for both named and regular constructors.

Which  makes the following lines completely redundant:

```
If the [=interface=] is declared with a [{{Constructor}}] [=extended attribute=],
then the [=interface object=] can be called as a constructor
to create an object that implements that interface.
Calling that interface as a function will throw an exception.

[=Interface objects=] whose [=interfaces=] are not declared
with a [{{Constructor}}] [=extended attribute=] will throw when called,
both as a function and as a constructor.
```

Maybe we should turn these into notes and let the algorithm speak for itself?

**Comment by domenic** [comment]
Hmm, why do you think it's implicit? What phrasing do you think triggers that ES clause? I still think to trigger that ES stuff you need to say something like

> Interface objects are [built-in function objects]\(link to ES spec\) that are constructors, and whose [[Call]] and [[Construct]] behavior is governed by the following algorithm:

and similarly for named constructors. Maybe the "that are constructors" is confusing to the casual reader, in which case I would add something like:

> Note: interface objects "are constructors" in the strict sense of the ECMAScript specification, that [IsConstructor] returns true when applied to them because they have a [[Construct]] internal method. But unless they are annotated with the [Constructor] extended attribute, attempting to construct an interface object will always throw. Thus they are often not "constructors" in the sense an ECMAScript programmer might expect.

I agree those lines are redundant from a specification perspective; I always thought of them as stating a fact, not a normative requirement (no "must", instead "will" or "can"). Moving them to a note seems fine; I'm never sure when it's better to put non-normative statements of fact inside notes or just leave them there.

Now that I re-read the above, I guess maybe my suggestions re-state a lot of those two lines you quote, although IMO in a more explicit way.

**Comment by Ms2ger** [comment]
Was this fixed in 4a8f57e2 by using `CreateBuiltinFunction()`?

**Comment by bzbarsky** [comment]
Looks like it.
